compiler/luci-interpreter/src/kernels/LogSoftmax.test.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
+ * Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kernels/LogSoftmax.h"
+#include "kernels/TestUtils.h"
+
+namespace luci_interpreter
+{
+namespace kernels
+{
+namespace
+{
+
+using namespace testing;
+
+TEST(LogSoftmaxTest, Float)
+{
+  Shape input_shape{2, 4};
+  std::vector<float> input_data{
+      0, -6, 2,  4, //
+      3, -2, 10, 1, //
+  };
+  Tensor input_tensor = makeInputTensor<DataType::FLOAT32>(input_shape, input_data);
+  Tensor output_tensor = makeOutputTensor(DataType::FLOAT32);
+
+  LogSoftmax kernel(&input_tensor, &output_tensor);
+  kernel.configure();
+  kernel.execute();
+
+  std::vector<float> ref_output_data{
+      -4.14297, -10.14297, -2.14297,   -.142971, //
+      -7.00104, -12.00104, -.00104087, -9.00104, //
+  };
+  EXPECT_THAT(extractTensorData<float>(output_tensor),
+              ElementsAreArray(ArrayFloatNear(ref_output_data)));
+}
+
+TEST(LogSoftmaxTest, Uint8)
+{
+  float kMin = -10;
+  float kMax = 10;
+  float kLogSoftmaxQuantizedTolerance = (kMax - kMin) / 255.0;
```

Comment:
I think this should be 16./256, the scale of output tensor.

Reply:
<pre><code>
Value of: dequantize<uint8_t>(extractTensorData<uint8_t>(output_tensor), output_tensor.scale(), output_tensor.zero_point())
Expected: has 8 elements where
element #0 is approximately -4.1429701 (absolute error <= 0.0625),
element #1 is approximately -10.14297 (absolute error <= 0.0625),
element #2 is approximately -2.1429701 (absolute error <= 0.0625),
element #3 is approximately -0.14297099 (absolute error <= 0.0625),
element #4 is approximately -7.00104 (absolute error <= 0.0625),
element #5 is approximately -12.00104 (absolute error <= 0.0625),
element #6 is approximately -0.00104087 (absolute error <= 0.0625),
element #7 is approximately -9.0010405 (absolute error <= 0.0625)
  Actual: { -4.125, -10.125, -2.125, -0.125, -7, -11.9375, 0, -8.9375 }, whose element #5 doesn't match, which is 0.0635405 from -12.001
/home/bearpaek/data/git/opensource/ONE/compiler/luci-interpreter/src/kernels/LogSoftmax.test.cpp:82: Failure
Value of: extractTensorData<uint8_t>(output_tensor)
Expected: has 8 elements where
element #0 is equal to '\xBD' (189),
element #1 is equal to ']' (93, 0x5D),
element #2 is equal to '\xDD' (221),
element #3 is equal to '\xFD' (253),
element #4 is equal to '\x8E' (142),
element #5 is equal to '?' (63, 0x3F),
element #6 is equal to '\xFF' (255),
element #7 is equal to 'o' (111, 0x6F)
  Actual: { '\xBD' (189), ']' (93, 0x5D), '\xDD' (221), '\xFD' (253), '\x8F' (143), '@' (64, 0x40), '\xFF' (255), 'p' (112, 0x70) }, whose element #4 doesn't match
[  FAILED  ] LogSoftmaxTest.Uint8 (0 ms)
</code></pre>

Test fails by 0.00104, when I change `kLogSoftmaxQuantizedTolerance = 16. / 256`.
I should look at where to fix...

Thank you for your advice, @jinevening :)

Reply:
P.S. Editted test source code
<details><summary><b>Source Code</b></summary>
<pre><code>
TEST(LogSoftmaxTest, Uint8)
{
  float kMin = -10;
  float kMax = 10;
  float kLogSoftmaxQuantizedTolerance = 16. / 256;
  std::pair<float, int32_t> quant_param = quantizationParams<uint8_t>(kMin, kMax);
  std::vector<float> input_data{
      0, -6, 2,  4, //
      3, -2, 10, 1, //
  };
  Tensor input_tensor{DataType::U8, {2, 4}, {{quant_param.first}, {quant_param.second}}, ""};
  Tensor output_tensor = makeOutputTensor(DataType::U8, 16. / 256, 255);
  std::vector<uint8_t> quantize_input =
      quantize<uint8_t>(input_data, quant_param.first, quant_param.second);
  input_tensor.writeData(quantize_input.data(), quantize_input.size() * sizeof(uint8_t));

  LogSoftmax kernel(&input_tensor, &output_tensor);
  kernel.configure();
  kernel.execute();

  std::vector<float> ref_output_data{
      -4.14297, -10.14297, -2.14297,   -.142971, //
      -7.00104, -12.00104, -.00104087, -9.00104, //
  };
  std::vector<int32_t> ref_output_shape{2, 4};
  EXPECT_THAT(dequantize<uint8_t>(extractTensorData<uint8_t>(output_tensor), output_tensor.scale(),
                                  output_tensor.zero_point()),
              ElementsAreArray(ArrayFloatNear(ref_output_data, kLogSoftmaxQuantizedTolerance)));
  EXPECT_THAT(extractTensorShape(output_tensor), ::testing::ElementsAreArray(ref_output_shape));
  EXPECT_THAT(extractTensorData<uint8_t>(output_tensor),
              ::testing::ElementsAreArray({189, 93, 221, 253, 142, 63, 255, 111}));
}
</code></pre></details>

Reply:
Just I Thought, Input data quantization make a more error on this. so, I think chage kMin to -16 and kMax to 16 passes test.
but, I'm not sure this change is enough.

Reply:
> Just I Thought, Input data quantization make a more error on this. so, I think changing kMin to -16 and kMax to 16 passes test. but, I'm not sure this change is enough.

<details><summary><b>Output</b></summary><pre><code>
/home/bear/data/opensource/ONE/compiler/luci-interpreter/src/kernels/LogSoftmax.test.cpp:81: Failure
Value of: extractTensorData<uint8_t>(output_tensor)
Expected: has 8 elements where
element #0 is equal to '\xBD' (189),
element #1 is equal to ']' (93, 0x5D),
element #2 is equal to '\xDD' (221),
element #3 is equal to '\xFD' (253),
element #4 is equal to '\x8E' (142),
element #5 is equal to '?' (63, 0x3F),
element #6 is equal to '\xFF' (255),
element #7 is equal to 'o' (111, 0x6F)
  Actual: { '\xBC' (188), '\\' (92, 0x5C), '\xDD' (221), '\xFD' (253), '\x8F' (143), '>' (62, 0x3E), '\xFF' (255), 'n' (110, 0x6E) }, whose element #0 doesn't match
</code></pre></details>

It solves tolerance problem, but still has problem with UINT8-typed reference output data.

Thank you for your advice!!!

Reply:
Oh, sorry that comment is mistake. sorry for confusing. On TFlite Example, the value you use is right.

Reply:
The error results from double precision and single precision. TF test code uses single precision (float) when computing zero point, while our quantize function uses double precision (double). The small difference is amplified by `round` function, causing a different zero point.
(In this test case, the zero point of input is 127 in TF but 128 in our code)

TF test code using float: 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/test_util.h#L725-L734

Our test code using double:
https://github.com/Samsung/ONE/blob/master/compiler/luci-interpreter/src/kernels/TestUtils.h#L134-L141

Using double precision is obviously more accurate, but we had better follow the TFLite codes since we are using test data from TFLite.

@struss , Can you change our code to use float instead of double?

@underflow101 , Please wait until our code is changed to use float. The test will pass afterwards.

Reply:
@jinevening okay, I will make

Reply:
I'd just fix the reference data though :man_shrugging:

Reply:
> I'd just fix the reference data though 

This can be a solution, but changing the quantization function may require less efforts if we are going to use TF test data in the future.

Reply:
> > I'd just fix the reference data though
> 
> This can be a solution, but changing the quantization function may require less efforts if we are going to use TF test data in the future.

This makes sense

Reply:
@underflow101 After #4076 merged, please rebase it on master.

Reply:
> The error results from double precision and single precision. TF test code uses single precision (float) when computing zero point, while our quantize function uses double precision (double). The small difference is amplified by `round` function, causing a different zero point.
> (In this test case, the zero point of input is 127 in TF but 128 in our code)
> 
> TF test code using float:
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/test_util.h#L725-L734
> 
> Our test code using double:
> https://github.com/Samsung/ONE/blob/master/compiler/luci-interpreter/src/kernels/TestUtils.h#L134-L141
> 
> Using double precision is obviously more accurate, but we had better follow the TFLite codes since we are using test data from TFLite.
> 
> @struss , Can you change our code to use float instead of double?
> 
> @underflow101 , Please wait until our code is changed to use float. The test will pass afterwards.

Thank you for your kind and detailed reply!!
I'll wait until our test codes to use float.

Reply:
> @underflow101 After #4076 merged, please rebase it on master.

Thank you, @struss!

Reply:
I rebased my fixed code on master.
Please take another look :)

Thank you!